src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -160,24 +160,44 @@ private Optional<ThingAssociations> getThingsAssociatedWithCoreDevice() {
         return Optional.empty();
     }
 
-    private void refresh(ThingAssociations associations) {
-        associations.getThings().forEach(this::refreshCertificateThingAttachment);
-        associations.getStaleThings().forEach(thingRegistry::deleteThing);
-
-        certificateRegistry.getAllCertificates()
-                .forEach(certificate -> {
-            if (associations.isAttachedToThing(certificate)) {
-                refreshCertificateValidity(certificate);
-            } else {
-                certificateRegistry.deleteCertificate(certificate);
-            }
-        });
+    private void refresh(Stream<Thing> thingsAttachedToCore) {
+        Pair<Set<String>, Consumer<Thing>> refreshCertificates = refreshCertificateValidity();
+
+        Set<String> attachedThingNames = thingsAttachedToCore
+                // Refresh the thing certificate attachments
+                .peek(this::refreshCertificateAttachments)
+                // Load the thing from the local store after their certificates have been updated.
+                .map(thing -> thingRegistry.getThing(thing.getThingName()))
+                // Skip the things that could not be found or for which the refresh failed
+                .filter(Objects::nonNull)
+                // Refresh the validity of all the certificates attached to the things
+                .peek(refreshCertificates.getRight())
+                // Get the names of the valid things that should still be in the registries
+                .map(Thing::getThingName)
+                // Wait for the stream to finish processing, collect all the valid attached things
+                .collect(Collectors.toSet());
+
+        // Clean up the registries by providing the names of the things that should still be attached and
+        // the ids of the certificates that are still attached to a thing.
+        cleanUpRegistries(attachedThingNames, refreshCertificates.getLeft());
     }
 
-    private void refreshCertificateThingAttachment(Thing thing) {
-        VerifyThingAttachedToCertificate useCase = useCases.get(VerifyThingAttachedToCertificate.class);
+    private void refreshCertificateAttachments(Thing thing) {
+        AtomicReference<Thing> thingWithCerts = new AtomicReference<>(thing);
+
+        if (thingWithCerts.get().getSource() != Thing.Source.LOCAL) {
+            thingWithCerts.set(thingRegistry.getThing(thing.getThingName()));
+        }
+
+        if (thingWithCerts.get() == null) {
+            logger.atInfo().kv("thingName", thing.getThingName())
+                    .log("No local version found for thing. Not refreshing thing certificate attachments");
+            return;
+        }
```

Comment:
this null-check should be before L#188 checking the source of this thing.

Reply:
Good catch